client/worker.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2019 Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	jaegerpb "github.com/jaegertracing/jaeger/model"
+	ocTrancelator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+	"go.opencensus.io/trace"
+
+	sapmpb "github.com/signalfx/sapm-proto/gen"
+)
+
+const (
+	defaultRateLimitingBackoffSeconds = 8
+	headerAuthToken                   = "X-SF-Token"
+	headerRetryAfter                  = "Retry-After"
+	headerContentEncoding             = "Content-Encoding"
+	headerContentType                 = "Content-Type"
+	headerValueGZIP                   = "gzip"
+	headerValueXProtobuf              = "application/x-protobuf"
+)
+
+type worker struct {
+	authToken  string
+	client     *http.Client
+	endpoint   string
+	gzipWriter *gzip.Writer
+	maxRetries uint64
+	doneCh     chan struct{}
+}
+
+func newWorker(client *http.Client, endpoint string, authToken string, maxRetries uint64) *worker {
+	w := &worker{
+		authToken:  authToken,
+		client:     client,
+		endpoint:   endpoint,
+		gzipWriter: gzip.NewWriter(nil),
+		maxRetries: maxRetries,
+		doneCh:     make(chan struct{}),
+	}
+	return w
+}
+
+func (w *worker) export(ctx context.Context, batch *jaegerpb.Batch) error {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	if len(batch.Spans) == 0 {
+		span.AddAttributes(trace.BoolAttribute("empty_batch", true))
+		return nil
+	}
+
+	sr, err := w.prepare(ctx, batch)
+	if err != nil {
+		recordEncodingFailure(ctx, sr)
+		span.SetStatus(trace.Status{
+			Code: trace.StatusCodeInvalidArgument,
+		})
+		return err
+	}
+
+	retries := uint64(0)
+
+	for {
+		err := w.send(ctx, sr)
+		if err == nil {
+			recordSuccess(ctx, sr)
+			break
+		}
+		recordSendFailure(ctx, sr)
+
+		if err.Permanent {
+			span.SetStatus(trace.Status{
+				Code: ocTrancelator.OCStatusCodeFromHTTP(int32(err.StatusCode)),
+			})
+			recordDrops(ctx, sr)
+			return err
+		}
+
+		if err.RetryDelaySeconds > 0 {
+			if retries >= w.maxRetries {
+				span.SetStatus(trace.Status{
+					Code: trace.StatusCodeDeadlineExceeded,
+				})
+				recordDrops(ctx, sr)
+				err.Permanent = true
+				return err
+			}
+			time.Sleep(time.Duration(err.RetryDelaySeconds) * time.Second)
```

Comment:
Maybe we could utilize a cancelable context here?

Reply:
We could but generally context is used for actual cancellation. I mean if the collector wanted to tell an exporter to cancel a particular request, it would cancel that request's context. The implication is that requests would be cancelled immediately instead of waiting to be finished first. So I'm not sure if we should mix this case with per request cancellations.